File: PAT_Basic_Level_by_cpp/1120 买地攻略/1120 买地攻略.cpp
// 1120 买地攻略
//
// 实现原理：
// 给定 N 块连续土地的价格和客户手中的现金量 M，客户只能购买连续相邻的土地。
// 求有多少种不同的购买方案，使得所选连续土地的总价不超过 M。
//
// 关键步骤：
// 1. 使用滑动窗口（双指针）：维护窗口 [left, right] 的和 sum。
// 2. 右指针 right 逐步右移，累加价格；当 sum > M 时，左指针 left 右移并减去价格。
// 3. 对每个 right，以 right 结尾且和不超过 M 的连续子数组个数为 (right - left + 1)，
//    累加到答案。
//
// 复杂度分析：
// 时间复杂度：O(N)，每个元素最多进出窗口一次。
// 空间复杂度：O(N)，存储价格数组。
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int n;
    long long m;
    if (!(cin >> n >> m)) return 0;
    vector<long long> a(n);
    for (int i = 0; i < n; ++i) cin >> a[i];

    long long cnt = 0, sum = 0;
    int left = 0;
    for (int right = 0; right < n; ++right) {
        sum += a[right];
        while (sum > m) { sum -= a[left]; ++left; }  // 收缩窗口使和不超过 M
        cnt += (right - left + 1);                   // 以 right 结尾的合法方案数
    }
    cout << cnt << endl;
    return 0;
}

File: PAT_Basic_Level_by_go/1120 买地攻略/1120 买地攻略.go
/*
题目：1120 买地攻略

解题思路：
给定 N 块连续的土地价格和预算 M，求有多少种连续子区间（即连续地块）
的价格之和不超过 M。
使用滑动窗口（双指针）法，右指针向右扩展，当和超过 M 时左指针右移，
每次右指针位置固定时，以 right 结尾的合法子区间数为 right - left + 1。
*/

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	// 初始化标准输入输出
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, m int
	fmt.Fscan(in, &n, &m)

	// 读取每块土地的价格
	prices := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &prices[i])
	}

	// 滑动窗口统计合法子区间数
	ans := 0
	left := 0
	sum := 0
	for right := 0; right < n; right++ {
		sum += prices[right]
		// 当窗口和超过 M 时，收缩左边界
		for sum > m {
			sum -= prices[left]
			left++
		}
		// 以 right 结尾的合法子区间数量
		ans += right - left + 1
	}
	fmt.Fprintln(out, ans)
}